onnxruntime/core/providers/cpu/reduction/reduction_ops.h

```diff
@@ -444,6 +454,96 @@ class ReduceAggregatorMax : public ReduceAggregator<T> {
+  }
+
+  static void FastReduceKR(const Tensor& input, const gsl::span<const int64_t>& fast_shape,
+                           Tensor& output, concurrency::ThreadPool* tp) {
+    const bool* data = input.Data<bool>();
+    bool* out = output.MutableData<bool>();
+    int64_t stridei = fast_shape[1];
+    concurrency::ThreadPool::TryParallelFor(
+        tp, onnxruntime::narrow<std::ptrdiff_t>(fast_shape[0]), ParallelReduceFastCost(1, stridei, sizeof(bool), 6),
+        [data, stridei, out](std::ptrdiff_t first, std::ptrdiff_t last) {
+          EigenVectorMap<bool>(out + first, last - first) = ConstEigenMatrixMap<bool>(
+                                                                data + first * stridei, onnxruntime::narrow<size_t>(stridei), last - first)
+                                                                .cast<int>()
+                                                                .colwise()
+                                                                .maxCoeff()
+                                                                .cast<bool>();
+        });
+  }
+
+  static void FastReduceRK(const Tensor& input, const gsl::span<const int64_t>& fast_shape,
+                           Tensor& output, concurrency::ThreadPool* tp) {
+    int64_t n_rows = fast_shape[0];
+    int64_t N = fast_shape[1];
+    const bool* data = input.Data<bool>();
+    bool* out = output.MutableData<bool>();
+    memcpy(out, data, SafeInt<size_t>(N) * sizeof(bool));
+
+    concurrency::ThreadPool::TryParallelFor(
+        tp, onnxruntime::narrow<std::ptrdiff_t>(N), ParallelReduceFastCost(1, n_rows, sizeof(bool), 6),
+        [data, out, N, n_rows](ptrdiff_t begin, ptrdiff_t end) {
+          const bool* p;
+          for (int64_t row = 1; row < n_rows; ++row) {
+            p = data + row * N;
+            for (int64_t j = begin; j < end; ++j) {
+              out[j] = out[j] || p[j];
+            }
+          }
+        });
+  }
+
+  static void FastReduceKRK(const Tensor& input, const gsl::span<const int64_t>& fast_shape,
+                            Tensor& output, concurrency::ThreadPool* tp) {
+    const bool* data = input.Data<bool>();
+    bool* out = output.MutableData<bool>();
+    int64_t stridei = fast_shape[1] * fast_shape[2];
+    int64_t strideo = fast_shape[2];
+    concurrency::ThreadPool::TryParallelFor(
+        tp, onnxruntime::narrow<std::ptrdiff_t>(fast_shape[0]), ParallelReduceFastCost(fast_shape[1], fast_shape[2], sizeof(bool), 6),
+        [data, fast_shape, stridei, strideo, out](ptrdiff_t begin, ptrdiff_t end) {
+          for (ptrdiff_t j = begin; j < end; ++j) {
+            EigenVectorMap<bool>(out + j * strideo, onnxruntime::narrow<size_t>(strideo)) =
+                ConstEigenMatrixMap<bool>(
+                    data + j * stridei, onnxruntime::narrow<size_t>(fast_shape[2]), onnxruntime::narrow<size_t>(fast_shape[1]))
+                    .cast<int>()
```

Comment:
Is there a way to avoid casting?

Reply:
The case is needed because maxCoeff does not support boolean type. So cases are needed.

Reply:
@xadupre as you suggested, the case target is now bool.